src/feat/feature-fbank.cc

```diff
@@ -61,125 +70,52 @@ const MelBanks *Fbank::GetMelBanks(BaseFloat vtln_warp) {
-  output->Resize(rows_out, cols_out);
-
-  // Optionally extract the remainder for further processing
-  if (wave_remainder != NULL)
-    ExtractWaveformRemainder(wave, opts_.frame_opts, wave_remainder);
-
-  // Buffers
-  Vector<BaseFloat> window;  // windowed waveform.
-  Vector<BaseFloat> mel_energies;
-  std::vector<BaseFloat> temp_buffer;  // used by srfft.
-  BaseFloat log_energy;
-
-  // Compute all the freames, r is frame index..
-  for (int32 r = 0; r < rows_out; r++) {
-    // Cut the window, apply window function
-    ExtractWindow(wave, r, opts_.frame_opts, feature_window_function_, &window,
-                  (opts_.use_energy && opts_.raw_energy ? &log_energy : NULL));
-
-    // Compute energy after window function (not the raw one)
-    if (opts_.use_energy && !opts_.raw_energy)
-      log_energy = Log(std::max(VecVec(window, window),
-                                std::numeric_limits<BaseFloat>::min()));
-
-    if (srfft_ != NULL)  // Compute FFT using split-radix algorithm.
-      srfft_->Compute(window.Data(), true, &temp_buffer);
-    else  // An alternative algorithm that works for non-powers-of-two.
-      RealFft(&window, true);
-
-    // Convert the FFT into a power spectrum.
-    ComputePowerSpectrum(&window);
-    SubVector<BaseFloat> power_spectrum(window, 0, window.Dim()/2 + 1);
-
-    // Sum with MelFiterbank over power spectrum
-    mel_banks.Compute(power_spectrum, &mel_energies);
-    if (opts_.use_log_fbank) {
-      // avoid log of zero (which should be prevented anyway by dithering).
-      mel_energies.ApplyFloor(std::numeric_limits<BaseFloat>::min());
-      mel_energies.ApplyLog();  // take the log.
-    }
+  // Convert the FFT into a power spectrum.
+  ComputePowerSpectrum(signal_frame);
+  SubVector<BaseFloat> power_spectrum(*signal_frame, 0,
+                                      signal_frame->Dim() / 2 + 1);
 
-    // Output buffers
-    SubVector<BaseFloat> this_output(output->Row(r));
-    SubVector<BaseFloat> this_fbank(this_output.Range((opts_.use_energy? 1 : 0),
-                                                      opts_.mel_opts.num_bins));
-
-    // Copy to output
-    this_fbank.CopyFromVec(mel_energies);
-    // Copy energy as first value
-    if (opts_.use_energy) {
-      if (opts_.energy_floor > 0.0 && log_energy < log_energy_floor_) {
-        log_energy = log_energy_floor_;
-      }
-      this_output(0) = log_energy;
-    }
+  int32 mel_offset = ((opts_.use_energy && !opts_.htk_compat) ? 1 : 0);
+  SubVector<BaseFloat> mel_energies(*feature,
+                                    mel_offset,
+                                    opts_.mel_opts.num_bins);
+
+  // Sum with mel fiterbanks over the power spectrum
+  mel_banks.Compute(power_spectrum, &mel_energies);
+  if (opts_.use_log_fbank) {
+    // avoid log of zero (which should be prevented anyway by dithering).
+    mel_energies.ApplyFloor(std::numeric_limits<BaseFloat>::epsilon());
+    mel_energies.ApplyLog();  // take the log.
+  }
```

Comment:
This pattern repeats often.Maybe add a new method, e. g. `ApplyFlooredLog(BaseFloat eps = 
std::numeric_limits<BaseFloat>::epsilon())`?